### PR TITLE
Docs: Add documentation from SDK testing session at the Singapore member meeting

### PR DESCRIPTION
**Pull request by cecille**
This is the markdown version of the material presented at the Singapore member meeting. Adding to the SDK to allow us to make updates more easily / for integration into the handbook if possible.

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

**Review by ReneJosefsen — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

File: docs/testing/index.md [hunk 1]
```diff
@@ -11,16 +11,33 @@ in the SDK.
...
 -   [YAML](./yaml.md)
 -   [Python testing framework](./python.md)
 -   [Enabling tests in the CI](./ci_testing.md)
+-   [Integration test utilities](./integration_test_utilities.md)
+
+## Unit testing
+
+Unit tests run on small pieces (“units”) of business logic. They do not use an
+external controller and instead test at the public interface of the class or
+function. For cluster, this requires an API that separates the cluster logic
```
Comment (rbultman) [inline docs/testing/index.md:34]:
```suggestion
function. For clusters, this requires an API that separates the cluster logic
```

File: docs/testing/unit_testing_clusters.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+### ClusterServer
+
+![](./img/unit_testable_clusters_server.png)
+
+The ClusterServerClass is a **Very** light wrapper over ClusterLogic. It
+translates Interaction Model wire format handling into API calls for cluster
+logic methods.
+
+This class iImplements both the AttributeAccessInterface and the CommandHandler
```
Comment (rbultman) [inline docs/testing/unit_testing_clusters.md:33]:
```suggestion
This class implements both the AttributeAccessInterface and the CommandHandler
```

File: docs/testing/unit_testing_clusters.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+interfaces so ClusterLogic properly handles data dependencies between command
```
Comment (rbultman) [inline docs/testing/unit_testing_clusters.md:34]:
```suggestion
interfaces so ClusterLogic properly handles data dependencies between commands
```

File: docs/testing/unit_testing_clusters.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+```
+
+### ClusterLogic
+
+![](./img/unit_testable_clusters_logic.png)
+
+The ClusterLogic class is for all the code that is SHARED between platforms. It
+does NOT include any TLV parsing or direct calls to Ember/IM/LogEvent etc.
+
+The class should include attribute getter/setters and handlers for all commands.
```
Comment (rbultman) [inline docs/testing/unit_testing_clusters.md:63]:
```suggestion
The class should include attribute getters/setters and handlers for all commands.
```

File: docs/testing/unit_testing_clusters.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+![](./img/unit_testable_clusters_context.png)
+
+### ClusterDriver
+
+The ClusterDriver is called by the ClusterLogic class and is used to translate
+attribute changes and commands into application actions. It also reports
+external changes back to the ClusterLogic class.
+
+The API design for this class with vary by the cluster, but it si generally
```
Comment (rbultman) [inline docs/testing/unit_testing_clusters.md:111]:
```suggestion
The API design for this class will vary by the cluster, but it is generally
```

File: docs/testing/unit_testing_clusters.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+recommended to use a generic API where possible, so the API ports easily to
+other platforms. For example an attribute changed callback with the changes
```
Comment (rbultman) [inline docs/testing/unit_testing_clusters.md:113]:
```suggestion
other platforms, for example an attribute changed callback with the changes
```

File: docs/testing/unit_testing_clusters.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+# Unit testing ClusterServer
+
+-   Best to have the lightest wrapping possible
+    -   If the wrapper is light, the code can be covered by integration or unit
+        tests, or a combination.
+    -   Correctness can mostly be validated by inspection if it’s trivial.
+-   Important tests
+    -   Errors when ClusterLogic instances aren’t properly registered.
+    -   Flow through to ClusterLogic for all reads/writes/command.
```
Comment (rbultman) [inline docs/testing/unit_testing_clusters.md:153]:
```suggestion
    -   Flow through to ClusterLogic for all reads/writes/commands.
```

**Review by ReneJosefsen — APPROVED** [review]
Assuming Robs comments gets committed 👍🏻

**Review by rbultman — APPROVED** [review]

**Review by robszewczyk — APPROVED** [review]

**Pull request merged**